### PR TITLE
#1315: Min/Median Block Mining

### PR DESCRIPTION
**Pull request by kantai**
## Description

This implements mean of min and median block mining #1315.

There's a couple things happening in this PR:

1. Identification of miners is done through UTXO chaining.
2. User burn operations may identify multiple block commits (the prior implementation relied on VRF key validation, which cannot be done before Stacks block reveal)
3. The burn distribution is calculated using a mean of min and median burns over a 6-block window.

## Type of Change
- New feature

## Does this introduce a breaking change?

Yes -- this is a consensus breaking change. Miners _do not_ need to change their behavior, though doing so would probably be advantageous.

## Testing information

There's unit test coverage for the new implementation of `BurnSamplePoint::make_min_median_distribution`. Integration with the sortdb / burnchain processing code is covered by existing integration tests (the integration itself is checked using panicking assertions in the new `BurnSamplePoint::make_min_median_distribution` function).

### CODE REVIEW

File: src/burnchains/burnchain.rs [hunk 1]
```diff
@@ -157,14 +160,55 @@ impl BurnchainStateTransition {
...
+
+        for blocks_back in 0..(MINING_COMMITMENT_WINDOW - 1) {
+            if parent_snapshot.block_height < (blocks_back as u64) {
+                debug!("Mining commitment window shortened because block height is less than window size";
+                       "block_height" => %parent_snapshot.block_height,
+                       "window_size" => %MINING_COMMITMENT_WINDOW);
+                break;
+            }
+            let block_height = parent_snapshot.block_height - (blocks_back as u64);
+            let sortition_id = match sort_tx.get_block_snapshot_by_height(block_height)? {
```
Comment (jcnelson) [inline src/burnchains/burnchain.rs:175]: It might be more efficient to just read the parent snapshot using the current snapshot's parent snapshot pointer (this would save you a MARF read per block height).

File: src/burnchains/burnchain.rs [hunk 2]
```diff
@@ -157,14 +160,55 @@ impl BurnchainStateTransition {
...
+                &sortition_id,
+            )?);
+        }
+
+        // reverse vecs so that windows are in ascending block height order
+        windowed_block_commits.reverse();
+        windowed_user_burns.reverse();
+
+        // figure out if the PoX sunset finished during the window
+        let window_end_height = parent_snapshot.block_height + 1;
```
Comment (jcnelson) [inline src/burnchains/burnchain.rs:194]: Does the window get disabled after PoX sunset?

File: src/chainstate/burn/db/sortdb.rs [hunk 3]
```diff
@@ -1424,14 +1430,47 @@ impl<'a> SortitionHandleConn<'a> {
...
+        ];
+        let shared_miners = query_count(self, qry, &args)? as u64;
 
-        query_rows(self, qry, &args)
+        assert!(
+            shared_miners >= 1,
+            "BUG: Should be at least 1 matching miner for the winning block commit"
+        );
+
+        for winning_user_burn in winning_user_burns.iter_mut() {
```
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:1469]: I would prefer it if this adjustment happened within the sortition distribution logic, instead of in the sortition DB.

File: src/burnchains/burnchain.rs (hunk 1 above)
Comment (kantai) [reply src/burnchains/burnchain.rs:175]: The snapshot only stores a burnchain header hash, which would need a MARF lookup to resolve to a SortitionId. We could extend the BlockSnapshot structure to store a parent pointer as SortitionId, though.

File: src/burnchains/burnchain.rs (hunk 2 above)
Comment (kantai) [reply src/burnchains/burnchain.rs:194]: No -- it just changes which output it tries to match in the BlockCommitOp for UTXO chaining (because after the sunset completes, the BlockCommitOp has 1 fewer output).

File: src/chainstate/burn/distribution.rs [hunk 4]
```diff
@@ -340,6 +442,217 @@ mod tests {
...
+        assert_eq!(result[0].candidate.txid, commits[5][0].txid);
+        assert_eq!(result[1].candidate.txid, commits[5][1].txid);
+
+        assert_eq!(result[0].user_burns.len(), 1);
+        assert_eq!(result[1].user_burns.len(), 1);
+
+        assert_eq!(result[1].user_burns[0].txid, user_burns[5][0].txid);
+        assert_eq!(result[0].user_burns[0].txid, user_burns[5][0].txid);
+    }
+
```
Comment (jcnelson) [inline src/chainstate/burn/distribution.rs:655]: Can you add a test for the edge case where the PoX sunset takes effect?

File: src/chainstate/burn/distribution.rs [hunk 5]
```diff
@@ -38,195 +39,301 @@ use util::log;
...
-    pub fn make_distribution(
-        all_block_candidates: Vec<LeaderBlockCommitOp>,
-        consumed_leader_keys: Vec<LeaderKeyRegisterOp>,
-        user_burns: Vec<UserBurnSupportOp>,
+    ///
+    /// * `block_commits`: this is a mapping from relative block_height to the block
+    ///     commits that occurred at that height. These relative block heights start
+    ///     at 0 and increment towards the present. When the mining window is 6, the
+    ///     "current" sortition's block commits would be in index 5.
+    /// * `block_commits`: this is a mapping from relative block_height to the user
```
Comment (jcnelson) [inline src/chainstate/burn/distribution.rs:84]: Typo -- `user_burns`?

File: src/chainstate/burn/distribution.rs [hunk 6]
```diff
@@ -38,195 +39,301 @@ use util::log;
...
-
-        // sanity check -- each block commit must refer to a consumed key
-        for i in 0..all_block_candidates.len() {
-            let bc = &all_block_candidates[i];
-            match key_index.get(&(bc.key_block_ptr as u64, bc.key_vtxindex as u32)) {
-                None => {
-                    panic!(
-                        "No leader key for block commitment {} at ({},{}) -- points to ({},{})",
-                        &bc.txid, bc.block_height, bc.vtxindex, bc.key_block_ptr, bc.key_vtxindex
+        // sanity check
```
Comment (jcnelson) [inline src/chainstate/burn/distribution.rs:95]: For clarity's sake, it might be worth putting the sanity check into its own private method.

File: src/chainstate/burn/db/sortdb.rs (hunk 3 above)
Comment (kantai) [reply src/chainstate/burn/db/sortdb.rs:1469]: This method is only ever invoked by the anchor block processing -- I could update this to return the original block commit ops and the number of block commits that were identified by the same user burns. Then the caller (`preprocess_anchor_block`) would be responsible for applying the adjustment before storing the staging user burns. Otherwise, I think this would entail a lot of refactoring of how the miner rewards are calculated.

File: src/chainstate/burn/distribution.rs [hunk 7]
```diff
@@ -38,195 +39,301 @@ use util::log;
...
+
+                let user_burn_target_key = UserBurnIdentifier {
+                    rel_block_height: rel_block_height as u8,
+                    key_vtxindex: key_vtxindex,
+                    key_block_ptr: key_block_ptr,
+                    block_hash: block_header_hash_160,
+                };
+
+                if let Some(user_burn_recipients) = user_burn_targets.get(&user_burn_target_key) {
+                    let per_recipient = burn_fee / (user_burn_recipients.len() as u64);
```
Comment (jcnelson) [inline src/chainstate/burn/distribution.rs:235]: Wait, didn't we already do this in the sortition DB?

File: src/chainstate/burn/distribution.rs [hunk 8]
```diff
@@ -38,195 +39,301 @@ use util::log;
...
+            .map(|mut linked_commits| {
+                let mut all_burns: Vec<_> = linked_commits
+                    .iter()
+                    .map(|commit| {
+                        if let Some(commit) = commit {
+                            (commit.op.burn_fee as u128) + (commit.user_burns as u128)
+                        } else {
+                            // use 1 as the linked commit min. this gives a miner a _small_
+                            //  chance of winning a block even if they haven't performed chained utxos yet
+                            1
```
Comment (jcnelson) [inline src/chainstate/burn/distribution.rs:289]: What happens if this is 0?

File: src/chainstate/burn/distribution.rs (hunk 7 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs:235]: No -- the function in the sortition db is only ever invoked by the block preprocessing logic. The primary difference between what's happening in that code and what's happening here is that in the other function, the caller wants specifically only the user burns that supported a particular known-to-have-won block. That data is used in matured rewards calculations.

Here, we need to evaluate all the possible user burns over the window.

File: src/chainstate/burn/distribution.rs [hunk 9]
```diff
@@ -38,195 +39,301 @@ use util::log;
...
-                        candidate: bc.clone(),
-                        key: consumed_leader_keys[*i].clone(),
-                        user_burns: vec![],
+        let window_size = block_commits.len() as u8;
+
+        // first, let's link all of the current block commits to the priors
+        let mut commits_with_priors: Vec<_> =
+            // start with the most recent
+            block_commits
+            .remove((window_size - 1) as usize)
```
Comment (jcnelson) [inline src/chainstate/burn/distribution.rs:132]: I'm a little confused by this -- are we dropping the current sortition's block commit here?

File: src/chainstate/burn/distribution.rs (hunk 8 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs:289]: It leads to a corner case where if there's no mining for some period of time, no miner will get pick for 3-5 blocks once mining resumes.

File: src/chainstate/burn/distribution.rs (hunk 9 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs:132]: No -- it's pulling it out of the input array to begin building an array with all the linked commits attached.

File: src/chainstate/burn/distribution.rs (hunk 7 above)
Comment (jcnelson) [reply src/chainstate/burn/distribution.rs:235]: ah, gotcha

File: src/burnchains/burnchain.rs (hunk 1 above)
Comment (jcnelson) [reply src/burnchains/burnchain.rs:175]: I thought the BlockSnapshot already stored the sortition ID?  If not, then don't worry about it in this PR.

File: src/chainstate/burn/distribution.rs (hunk 4 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs:655]: Yep, will add

File: src/chainstate/burn/db/sortdb.rs (hunk 3 above)
Comment (jcnelson) [reply src/chainstate/burn/db/sortdb.rs:1469]: Ah, I see.

**Review by jcnelson — APPROVED** [review]
This LGTM, but maybe just add one more test to cover the case where PoX sunset occurs during the calculation.

File: src/chainstate/burn/distribution.rs (hunk 6 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs]: Updated in 5c80770d4

File: src/chainstate/burn/distribution.rs (hunk 5 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs]: Updated in 5c80770d4

File: src/chainstate/burn/distribution.rs (hunk 4 above)
Comment (kantai) [reply src/chainstate/burn/distribution.rs:764]: Added in 5c80770d4

**Review by lgalabru — APPROVED** [review]
Looks good to me, thank you @kantai!
I keep wondering if the rigidity of our expectations, when diving into the outputs of a BTC transaction, is a good thing or not (the order, the total count, etc).
I understand that it simplifies the retrieval, but there are 2 downsides that I can think about:
- Combining Stacks mining / support with BTC transactions: transaction's price are rising, BTC fee costs of Stacks mining could be amortized if the transaction could be used for something else.   
- Combining / Multiplexing Stacks protocols with other protocols: can we / what kind of meta-protocols can be built by combining some Bitcoin protocols (Stacks, Lightning, wBTC, etc).

**Comment by kantai** [comment]
Opened an issue to discuss the rigidity of our bitcoin wire formats https://github.com/blockstack/stacks-blockchain/issues/2094
